### DISCUSSION TITLE
Use case: in-memory repository with references to on-disk database

### DISCUSSION THREAD

**Discussion by jeffparsons** [GitoxideLabs/gitoxide#124] [Ideas]
Hi there! :wave: 

I was wondering whether you had any existing plan to support a use case that interests me: in-memory Git repositories. I'm specifically interested in using an in-memory repository to make new commits (or just tree objects), but referring to an existing on-disk repository for all older content.

The context for this is a tool where it would be most convenient for me to make "speculative" commits without writing them to disk and then only have to deal with Git trees, rather than having to deal with both Git trees _and_ uncommitted on-disk changes.

My workaround for this would be to have the tool create a "RAM disk" and make a new repository there for the speculative commits.

Thanks!

**Comment by Byron** [comment]
Hi,

thanks for starting this conversation!

Even though there aren't many tools writing objects yet nor facilities to build trees from indices or other sources, there is a Write trait <https://github.com/Byron/gitoxide/blob/8b1043483cb46fd1b7f47a90c9dce24a65d58d1b/git-odb/src/write.rs#L5> which they all accept. It's implemented for loose objects or the sink, a /dev/null for objects. Based on the latter you could easily implement an in-memory object store.

Creating commits on the other hand is nothing gitoxide can help you with, yet, and it's probably further away given the current roadmap.

When thinking of the RAM disk approach I believe using git-alternates would be the way to go. Writes go into the RAM disk and reads are falling through to the original repository.

I hope that helps.

**Reply by jeffparsons** [reply]
Thanks for your response.

Looking through the _git-object_ documentation, I think there might be enough already present to achieve what I want.

Whenever I finally get around to actually implementing this, I'll share what I come up with. :)
